### ISSUE TITLE
How to generate noisy_npy for testing??

### ISSUE THREAD

**Issue by edwardcho** [Magauiya/Extended_SURE#3]
I am testing your code using my data.
I ran train-code and got model.
After training, I tried inference-code (test-code).
But, I did not have sigma25 (or sigma50) npy files for testing.

How to generate this files??
(I think that this files don't have inference-time. Right?
 If so, this files should generate automatically at inference-time.)

**Comment by Magauiya** [comment]
Hi! 
I made a wrong file path in the code. Today (in 12 hours), I will correct all path mistakes. For now, you can change it by yourself. All test files including sigma25 are here: [BSD68]{https://github.com/Magauiya/Extended_SURE/tree/master/Dataset/testset/CBSD68} and [CSet9]{https://github.com/Magauiya/Extended_SURE/tree/master/Dataset/testset/CSet9}

**Comment by Magauiya** [comment]
> I am testing your code using my data.
> I ran train-code and got model.
> After training, I tried inference-code (test-code).
> But, I did not have sigma25 (or sigma50) npy files for testing.
> 
> How to generate this files??
> (I think that this files don't have inference-time. Right?
> If so, this files should generate automatically at inference-time.)

**Comment by Magauiya** [comment]
> I am testing your code using my data.
> I ran train-code and got model.
> After training, I tried inference-code (test-code).
> But, I did not have sigma25 (or sigma50) npy files for testing.
> 
> How to generate this files??
> (I think that this files don't have inference-time. Right?
> If so, this files should generate automatically at inference-time.)

Hi!
Those files are located here: [BSD68](https://github.com/Magauiya/Extended_SURE/tree/master/Dataset/testset/CBSD68) and [CSet9](https://github.com/Magauiya/Extended_SURE/tree/master/Dataset/testset/CSet9). You can explicitly give those paths for now. Sorry about that.

I made a mistake while giving path in the code. So, I am gonna correct all paths, including the path for patch generation.

**Comment by edwardcho** [comment]
Hello Sir,
For inference of my own data, I should make noisy-npy data. Right??
If so, How do generate this file??  

![image](https://user-images.githubusercontent.com/29569662/74204375-3a0e5c80-4cb6-11ea-9487-1fcb42d33bb0.png)

**Comment by Magauiya** [comment]
Same as in generating noisy patches: `noisy = clean + np.random.normal(0, sigma, (H, W, 3), dtype="float32")` 

Then, use those noisy images.

**Comment by Magauiya** [comment]
> Hello Sir,
> For inference of my own data, I should make noisy-npy data. Right??
> If so, How do generate this file??
> 
> ![image](https://user-images.githubusercontent.com/29569662/74204375-3a0e5c80-4cb6-11ea-9487-1fcb42d33bb0.png)

Same as in generating noisy patches: `noisy = clean + np.random.normal(0, sigma, (H, W, 3), dtype="float32")` 

Then, use those noisy images. 
NOTE: if your network is trained on [0-1] image range, then divide your sigma to 255. For example **sigma = 25**, then the input sigma should be **0.098**

**Comment by Reeshad-Khan** [comment]
Hi Sir,
Whenever I try to train the network, I get NAN mse. Could you please reply to that?

**Comment by Magauiya** [comment]
> Hi Sir, Whenever I try to train the network, I get NAN mse. Could you please reply to that?

Hello Reeshad-Khan, can you please elaborate? What script did you run and under what configurations (loss function, sigma, dataset, etc.) you run the model?

**Comment by Reeshad-Khan** [comment]
Hi Sir,
I ran "main_blind_rgb.py" with configurations( Loss function: SURE; sigma value: 2.5, 5, 7.5, 10, 12.5, 15, 17.5, 20, 22.5 1nd 25; dataset: I used an MRI dataset with png extension). I have used the CDnCNN network.
During the training I got this:
Epoch: [ 1] [   1/ 775] Time: 3.6169 MSE: 0.000032, sure: 16.643593
![eval1_0](https://user-images.githubusercontent.com/65279631/190542003-c662b158-854c-4423-93a0-f02531c1aa54.png)

Epoch: [ 1] [ 101/ 775] Time: 17.8714 MSE: nan, sure: nan
Output:
![output_eval1_973](https://user-images.githubusercontent.com/65279631/190542296-4f2bc3f4-35b6-46ec-8a7c-68c0996c1b58.png)

and after finishing the training:
Epoch: [50] [ 726/ 775] Time: 5558.0455 MSE: nan, sure: nan
Output:
![output_eval5_37947](https://user-images.githubusercontent.com/65279631/190542371-7b5ba462-e802-4e15-941e-10e7898fbcc8.png)

**Comment by Magauiya** [comment]
It is weird that even MSE diverges during the training. My suggestion is to check loaded dataset. Regarding the SURE training, when you add Gaussian noise to the input image, you should not clip the image range. Can you please share your code part, where you add dataset or whole changes you have made?

P. S. We have tried SURE and eSURE for a vast range of image types from grayscale (MRI, CT, natural images) to color and haven`t noticed this kind of error. SURE and eSURE assumes only one thing: noise should be additive Gaussian.

**Comment by Reeshad-Khan** [comment]
Thanks for your reply.
For a better understanding of the problem, if it's okay could you please check this link(https://github.com/Reeshad-Khan/SURE.git)? I have uploaded the whole project there to share with you.

**Comment by Reeshad-Khan** [comment]
Hello Sir @Magauiya ,
I really don't know what I am doing wrong. This time I just ran your project without making any changes. Even I kept the data you used. But still, I am getting NAN MSE after 1 Epoch. Please note that I have removed the existing checkpoint folder before starting training. And that's what I got: 
![Screen Shot 2022-10-10 at 8 15 06 PM](https://user-images.githubusercontent.com/65279631/194976770-accca33d-1903-4e0f-9ba0-f132834c1472.png)
And the images are like that:
![eval4_100950](https://user-images.githubusercontent.com/65279631/194977007-736631e9-f220-42ad-8a56-335484c92edd.png)
Also, I have tried training without removing the existing checkpoint folder and the results are as follows after the training:
![Screen Shot 2022-10-10 at 8 29 59 PM](https://user-images.githubusercontent.com/65279631/194977249-adcf4153-c852-44a2-8e38-e6daa177e51f.png)
and
![eval4_100950](https://user-images.githubusercontent.com/65279631/194977266-a2d8c3ff-ada6-4ab2-8987-5240f860374d.png)
So could you please tell me what I am doing wrong? Thanks in advance.

**Comment by Magauiya** [comment]
Hello @Reeshad-Khan,

That is very weird to me. Possible reasons might be the tensorflow version (what TF are you using?) and python versions. I double checked everything, but haven`t found anything. It works fine for my case (Tensorflow 1.10.0 python 2.7) Have you tried to just inference with existing checkpoints using [command here](https://github.com/Magauiya/Extended_SURE/tree/1bcec3b61d7c3687f0fc906d684cc7d29e59aab4#testing-using-the-trained-network) to see if model adequately denoise images?

P.S. Next week I`m planning to release pytorch implementation. I guess this will be more convenient for the community.

Sincerely,
Magauiya

**Comment by Reeshad-Khan** [comment]
Dear @Magauiya 
Thanks for your reply.
Currently I am using following for this project:
Python 2.7
tensorboard 1.10.0                   
tensorflow  1.10.0 
That would be great if the pytorch implementation is here. Also,, if possible may I see the MRI dataset you have used and the input pipeline for MRI dataset? My email is: rk010@uark.edu if you need.

**Comment by Reeshad-Khan** [comment]
![eval2_100950](https://user-images.githubusercontent.com/65279631/195208548-1ffa0d32-0482-4532-be85-2b818675beaf.png)
This is what I get using the existing checkpoints

**Comment by Magauiya** [comment]
@Reeshad-Khan please check pytorch version of the MC-SURE. https://github.com/Magauiya/Extended_SURE/tree/pytorch_version

I am working on implementing extended MC-SURE and replicating the same results as in my paper.

**Comment by Reeshad-Khan** [comment]
> 

Excellent. Thank you very much. I will check it. Thanks again!
